### ISSUE TITLE
Declare type: simplify FQN automatically

### ISSUE THREAD

**Issue by rentalhost** [kalessil/phpinspectionsea#255]
Following #28, is important simplify automatically yet-imported classes.

**Code before declare type:**

```php
use \MyNamespace\AnotherClass;

    ...
    // Inspection: \MyNamespace\AnotherClass can be declared as return type hint.
    /** @return AnotherClass */
    public function getAnotherClass() {
    }
    ...
```

**After QF:**

```php
    ...
    /** @return AnotherClass */
    public function getAnotherClass(): \MyNamespace\AnotherClass {
    }
    ...
```

**Expected after QF:**

```php
    ...
    /** @return AnotherClass */
    public function getAnotherClass(): AnotherClass {
    }
    ...
```

---

**Semi-related**: the tooltip that is shown to current inspection is "': {class} can be...". [There are](https://cloud.githubusercontent.com/assets/421851/25088070/eedda97a-2349-11e7-926b-dcfdc2dd9271.png) a `':` before the class name.

**Comment by funivan** [comment]
It is interesting if we can check phpstorm options. In phpstorm we have option: `automatically import classes`

**Comment by kalessil** [comment]
Should be the same behaviour as for ::class constant usage QF, hence removing easy to pick.

**Comment by Garethp** [comment]
@kalessil I hope you don't mind that I've picked this up. I figured I already knew somewhat how teh `:class constant` QF worked, so it would be a good ticket for me to pick up and dig deeper

**Comment by kalessil** [comment]
Good idea and I have additional input: QF for `::class` and this one should have the same import settings (auto-import, relative FQNs).

The part in `::class` is more stable (and here is rather POC). I'd like the fixer from `::class` to be generalized (without decomposition) for using also in return type hints.

**Comment by Garethp** [comment]
That's the plan. I'll probably end up refactoring the `::class` inspection and QF entirely in the process, since 90% of the logic in there basically relates to what should be common functionality.

Can you please expand on what you mean by
> without decomposition, probably an abstract class with shared importing functionality

My current approach was to create a utility class that contained the following functions: `isClassImported`, `importClass` and possibly `resolveClass` (If the logic from the problem registering would be useful elsewhere). Would you prefer a different approach?

**Comment by kalessil** [comment]
Yes, my idea was doing this:

1) move the fixer into a separate class (we have fixers package) for ::class
2) add an abstract method responsible for the fixing reported expressions (base fixer still big and responsible for importing, but no utils introduced - minimal changes)
3) ::class, return type hints got small fixes injecting class references supplied by the base class
4) a new service class can be introduced for the importing functionality at this step

Over util classes, I'd suggest following a more declarative approach. Something like this:
```java
   /* inside QFer, dispatching booleans is still bad but acceptable as we are OSS */
   final ClassImportService importService = new  ClassImportService(boolean importAtomatically, boolean useRelativeQN);
   final String qualifierToUse =  importService.import(String qualifier, PsiFile file);
   if (null != qualifierToUse) { // null if importAtomatically is false
      this.delegateCodeModification(PsiElement reportedExpression, qualifierToUse);
   }
```

- the code fragment is the business logic, so needs to be readable
- delegateCodeModification is abstract and implemented by children
- ClassImportService will contain the logic which you moving to UtilClasses
  - the logic there is very low-level, no need to structure it
  - the logic here is not a business logic, no need to put additional efforts

Does it make sense?

**Comment by Garethp** [comment]
I think so. I'll give it a shot over the next couple of days, and I'll ping you on the PR when I think I've got something closer to what you want, and see if I went the right way

**Comment by rentalhost** [comment]
Closing: fixed on #320 

The import process is not required once that we can use the environment intention programatically (as "Import class" and "Simplify FQN").
